lib/dynamoid/adapter_plugin/aws_sdk_v2.rb

```diff
@@ -559,6 +569,15 @@ def scan(table_name, scan_hash, select_opts = {})
           record_count = 0
           scan_count = 0
           loop do
+            # Adjust the limit down if the remaining record or scan limit are
+            # lower to obey limits. We can assume the difference won't be
+            # negative due to break statements below.
+            if request[:limit] && record_limit && record_limit - record_count < request[:limit]
+              request[:limit] = record_limit - record_count
+            elsif request[:limit] && scan_limit && scan_limit - scan_count < request[:limit]
+              request[:limit] = scan_limit - scan_count
+            end
+
```

Comment:
> SCAN will only scan through maximum of 1MB of data

Hm, I don't see the statement you've quoted. Most statements can be interpreted as the resulting set has maximum of 1 MB. Nothing about the amount of data that is looked at in underlying query/scan.

I have a large query example in which there are 200 records generated and 64 records will be 1 MB in size.

When doing a query with scan limit of 100. The query will return `scanned_count` as 100. But when run with a scan, it returns `scanned_count` as 64. Which shows the discrepancy (but I've noticed there are discrepancies with the local dynamodb jar and how production actual works) 

## Record Limit

You are correct! I kept thinking that the limit past was the limit of returned records, this is **not** correct. The limit given to the SDK is the scanned limit (number of records dynamodb looks at when evaluating criteria).

I will add the note as you suggested. So we have the trade off of:

* Either make small sets of HTTP calls to get the last remaining records
* Do NOT scale down limit when `record_count` approaches `record_limit` and instead during enumeration, stop once we've hit the limit. This will thus load `limit - remaining_records_needed` extra records. So in your example we'd actually be fine. But if there were 1000 extra records satisfying the condition then we'd have 999 extra records loaded into memory.

I don't see any other way to solve this since you basically the in between would be requesting in smaller chunks but you're just moving the needle from calling fewer HTTP calls or more HTTP calls.